src/theory/term_registration_visitor.cpp

```diff
@@ -112,74 +106,96 @@ void PreRegisterVisitor::visit(TNode current, TNode parent) {
+}
+void PreRegisterVisitor::preRegisterWithTheory(TheoryEngine* te,
+                                               TheoryIdSet& visitedTheories,
+                                               TheoryId id,
+                                               TNode current,
+                                               TNode parent)
+{
+  if (TheoryIdSetUtil::setContains(id, visitedTheories))
   {
-    visitedTheories =
-        TheoryIdSetUtil::setInsert(parentTheoryId, visitedTheories);
-    d_visited[current] = visitedTheories;
-    Theory* th = d_engine->theoryOf(parentTheoryId);
-    th->preRegisterTerm(current);
-    Debug("register::internal") << "PreRegisterVisitor::visit(" << current << "," << parent << "): adding " << parentTheoryId << std::endl;
+    // already registered
+    return;
   }
-  if (useType) {
-    if (!TheoryIdSetUtil::setContains(typeTheoryId, visitedTheories))
+  if (Configuration::isAssertionBuild())
+  {
+    Debug("register::internal")
+        << "PreRegisterVisitor::visit(" << current << "," << parent
+        << "): adding " << id << std::endl;
+    // This should never throw an exception, since theories should be
+    // guaranteed to be initialized.
+    // These checks don't work with finite model finding, because it
+    // uses Rational constants to represent cardinality constraints,
+    // even though arithmetic isn't actually involved.
+    if (!options::finiteModelFind())
     {
-      visitedTheories =
-          TheoryIdSetUtil::setInsert(typeTheoryId, visitedTheories);
-      d_visited[current] = visitedTheories;
-      Theory* th = d_engine->theoryOf(typeTheoryId);
-      th->preRegisterTerm(current);
-      Debug("register::internal") << "PreRegisterVisitor::visit(" << current << "," << parent << "): adding " << parentTheoryId << std::endl;
+      if (!te->isTheoryEnabled(id))
+      {
+        const LogicInfo& l = te->getLogicInfo();
+        LogicInfo newLogicInfo = l.getUnlockedCopy();
+        newLogicInfo.enableTheory(id);
+        newLogicInfo.lock();
+        std::stringstream ss;
+        ss << "The logic was specified as " << l.getLogicString()
+           << ", which doesn't include " << id
+           << ", but found a term in that theory." << std::endl
+           << "You might want to extend your logic to "
+           << newLogicInfo.getLogicString() << std::endl;
+        throw LogicException(ss.str());
+      }
     }
   }
-  Debug("register::internal")
-      << "PreRegisterVisitor::visit(" << current << "," << parent
-      << "): now registered with "
-      << TheoryIdSetUtil::setToString(visitedTheories) << std::endl;
-
-  Assert(d_visited.find(current) != d_visited.end());
-  Assert(alreadyVisited(current, parent));
+  // call the theory's preRegisterTerm method
+  visitedTheories = TheoryIdSetUtil::setInsert(id, visitedTheories);
+  Theory* th = te->theoryOf(id);
+  th->preRegisterTerm(current);
 }
 
+void PreRegisterVisitor::start(TNode node) {}
```

Comment:
what is this for?
If this is supposed to be overriden always, maybe assert false?

Reply:
Its a slightly different style, this visitor is intended to instantiate a template type on NodeVisitor (https://github.com/CVC4/CVC4/blob/master/src/expr/node_visitor.h), which is why these methods need to be defined.

I'm not a big fan of this style, but I don't feel strongly about refactoring this further (yet).